### PR TITLE
ES6 Destructuring

### PR DESCRIPTION
**Pull request by webb04**
## TL;DR;

ES6 destructuring is a syntax modernization that improves readability and reduces the amount of code necessary for expressive declarations.
## Username

[@webb04](https://github.com/webb04)
## Extra

ES6 destructuring is a syntax modernization that improves readability and reduces the amount of code necessary for expressive declarations.
### Array destructuring

``` javascript
// access first three items in an Array
const someArray = [1,2,3];
const [first, second, third] = someArray;

// possible to skip over elements
const [,,third] = ["el1", "el2", "el3"];
console.log(third);
// "el3"

// capture the remaining elements using the ES6 "rest" pattern
const [head, body, ...tail] = [1, 2, 3, 4];
console.log(tail);
// [3, 4]

// nice way to swap variables
let first = 1;
let second = 2;
[first, second] = [second, first];

// nice way to return a tuple from a function
function tuple() {
  return [1, 2];
}

const [first, second] = tuple();
```
### Object destructuring

``` javascript
 // bind property to a constant
 const personA = { name: "foo" };
 const personB = { name: "bar" };

 const { name: nameA } = personA;
 const { name: nameB } = personB;

 console.log(nameA);
 // "foo"
 console.log(nameB);
 // "bar"

// shortcut for when property and variable names are the same
 let { foo, bar } = { foo: "first", bar: "second" };
 console.log(foo);
 // "first"
 console.log(bar);
 // "second"

 // you can specify default values
 const {firstName = "John", lastName: surname = "Doe"} = {};
```

You can find more information about JavaScript ES6 destructuring here:
- [Getting Started with JavaScript ES6 Destructuring](https://strongloop.com/strongblog/getting-started-with-javascript-es6-destructuring/)
- [ES6 In Depth: Destructuring](https://hacks.mozilla.org/2015/05/es6-in-depth-destructuring/)

### CODE REVIEW

**Comment by loverajoel** [comment]
@hsleonis  hey use the PR template -> https://github.com/loverajoel/jstips/blob/gh-pages/GIT_TEMPLATE.md

**Comment by webb04** [comment]
@loverajoel Hey hope that's ok now

**Comment by loverajoel** [comment]
@webb04 the PR template is for the PR message 
image
, and the Post template is for the post.

File: _posts/en/2016-xx-xx-Destructuring.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+// shortcut for when property and variable names are the same
+ let { foo, bar } = { foo: "first", bar: "second" };
+ console.log(foo);
+ // "first"
+ console.log(bar);
+ // "second"
+
+ // you can specify default values
+ const {firstName = "John", lastName: surname = "Doe"} = {};
```
Comment (gromgit) [inline _posts/en/2016-xx-xx-Destructuring.md:69]: Show values of `firstName` and `surname`, to illustrate the effects of default values.

File: _posts/en/2016-xx-xx-Destructuring.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+---
+
+ES6 destructuring is a syntax modernization that improves readability and reduces the amount of code necessary for expressive declarations.
+
+### Array destructuring
+
+```javascript
+// access first three items in an Array
+const someArray = [1,2,3];
+const [first, second, third] = someArray;
```
Comment (gromgit) [inline _posts/en/2016-xx-xx-Destructuring.md:21]: Show results here.

File: _posts/en/2016-xx-xx-Destructuring.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+// possible to skip over elements
+const [,,third] = ["el1", "el2", "el3"];
+console.log(third);
+// "el3"
+
+// capture the remaining elements using the ES6 "rest" pattern
+const [head, body, ...tail] = [1, 2, 3, 4];
+console.log(tail);
```
Comment (gromgit) [inline _posts/en/2016-xx-xx-Destructuring.md:30]: Also show values of `head` and `body` for completeness.

File: _posts/en/2016-xx-xx-Destructuring.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+// [3, 4]
+
+// nice way to swap variables
+let first = 1;
+let second = 2;
+[first, second] = [second, first];
```
Comment (gromgit) [inline _posts/en/2016-xx-xx-Destructuring.md:36]: Showing the values of `first` and `second` here are **especially** important, to prove the swap was actually done.

File: _posts/en/2016-xx-xx-Destructuring.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+// nice way to return a tuple from a function
+function tuple() {
+  return [1, 2];
+}
+
+const [first, second] = tuple();
```
Comment (gromgit) [inline _posts/en/2016-xx-xx-Destructuring.md:43]: Show results here.

**Comment by zenopopovici** [comment]
Closed due to inactivity.
